packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -370,6 +372,57 @@ function warnOnHookMismatchInDev(currentHookName: HookType): void {
   }
 }
 
+function warnIfAsyncClientComponent(
+  Component: Function,
+  componentDoesIncludeHooks: boolean,
+) {
+  if (__DEV__) {
+    // This dev-only check only works for detecting native async functions,
+    // not transpiled ones. There's also a prod check that we use to prevent
+    // async client components from crashing the app; the prod one works even
+    // for transpiled async functions. Neither mechanism is completely
+    // bulletproof but together they cover the most common cases.
+    const prototype: Object | null = Object.getPrototypeOf(Component);
+    if (prototype && prototype[Symbol.toStringTag] === 'AsyncFunction') {
```

Comment:
`Object.prototype.toString.call(Component) === '[object AsyncFunction]'` could maybe be sufficient too. A bit more old school way of checking but seems more in line with how people do brand checking today.